### PR TITLE
Capture ChainReorg, FinalizedCheckpoint, and NewHead Events in Beacon Node

### PR DESCRIPTION
**Pull request by rauljordan**
**What type of PR is this?**

> Feature

**What does this PR do? Why is it needed?**

As part of #9001, we need to capture a few important events in eth2 that are required for the eth/v1/events standard API endpoint. These include chain reorgs, finalized checkpoint, and new chain head.

**IMPORTANT**

The eth2.0-apis chain reorg event does not capture enough information about whether the reorg was across epoch boundaries. Currently, we have a `StreamDuties` endpoint that is **UNUSED** in Prysm which only resends duties if there is a reorg across epoch boundaries. Unfortunately, we can no longer tell this information and therefore we simply send the duties if there is a reorg at all. This should not be be very harmful as reorgs should be very rare and the endpoint is unused in v1alpha1.

### CODE REVIEW

File: beacon-chain/blockchain/head.go [hunk 1]
```diff
@@ -142,6 +156,50 @@ func (s *Service) saveHead(ctx context.Context, headRoot [32]byte) error {
...
+			Type: statefeed.NewHead,
+			Data: &ethpbv1.EventHead{
+				Slot:                      newHeadSlot,
+				Block:                     headRoot[:],
+				State:                     newStateRoot,
+				EpochTransition:           helpers.IsEpochEnd(newHeadSlot),
+				PreviousDutyDependentRoot: previousDutyDependentRoot,
+				CurrentDutyDependentRoot:  currentDutyDependentRoot,
+			},
+		})
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/head.go:200]: I commented this out locally and no test failed. Please add a test for this functionality.

**Comment by prestonvanloon** [comment]
Build failure
```
beacon-chain/rpc/validator/assignments_test.go:348:11: undefined: "github.com/prysmaticlabs/prysm/beacon-chain/core/feed/state".ReorgData
beacon-chain/rpc/validator/assignments_test.go:356:11: undefined: "github.com/prysmaticlabs/prysm/beacon-chain/core/feed/state".ReorgData
```

File: beacon-chain/core/feed/state/events.go [hunk 2]
```diff
@@ -55,11 +59,3 @@ type InitializedData struct {
...
 	GenesisValidatorsRoot []byte
 }
-
-// ReorgData is the data alongside a reorg event.
-type ReorgData struct {
-	// NewSlot is the slot of new state after the reorg.
-	NewSlot types.Slot
-	// OldSlot is the slot of the head state before the reorg.
-	OldSlot types.Slot
-}
```
Comment (prestonvanloon) [inline beacon-chain/core/feed/state/events.go:65]: This is failing builds.

beacon-chain/rpc/validator/assignments_test.go:348:11: undefined: "github.com/prysmaticlabs/prysm/beacon-chain/core/feed/state".ReorgData
beacon-chain/rpc/validator/assignments_test.go:356:11: undefined: "github.com/prysmaticlabs/prysm/beacon-chain/core/feed/state".ReorgData

**Review by nisdas — COMMENTED** [review]
This PR inserts a lot of new code into our critical processing path, have any benchmarks been done on if there are any effects here on block processing time?

File: beacon-chain/blockchain/head.go [hunk 3]
```diff
@@ -290,3 +310,53 @@ func (s *Service) getJustifiedBalances() []uint64 {
 	defer s.justifiedBalancesLock.RUnlock()
 	return s.justifiedBalances
 }
+
+// Notifies a common event feed of a new chain head event. Called right after a new
+// chain head is determined, set, and saved to disk.
+func (s *Service) notifyNewHeadEvent(
+	newHeadSlot types.Slot,
```
Comment (terencechain) [inline beacon-chain/blockchain/head.go:317]: Optional feedback:
These arguments could be cleaner and reduced...  (ie. get slot from block, get state root from block... etc)

**Review by terencechain — APPROVED** [review]

File: beacon-chain/blockchain/head.go (hunk 3 above)
Comment (rauljordan) [reply beacon-chain/blockchain/head.go:317]: I can fix in the next PR which integrates this one

**Pull request merged**